capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -0,0 +1,177 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.sps.data.VolumeData;
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Collection;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns search data from the Google Books API. */
+@WebServlet("/api/search")
+public class SearchServlet extends HttpServlet {
+
+  public final static int DEFAULT_MAX_RESULTS = 5;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String fullOutput = "";
+    String searchTerm = request.getParameter("searchTerm");
+    if (searchTerm == null || searchTerm.isEmpty()) return;
+    searchTerm = URLEncoder.encode(searchTerm, "UTF-8");
+
+    int maxResults = parseNaturalNumber(request.getParameter("maxResults"));
+    if (request.getParameter("maxResults") != null) { 
+      maxResults = parseNaturalNumber(request.getParameter("maxResults"));
+    } 
+    if (maxResults == -1) {
+      maxResults = DEFAULT_MAX_RESULTS;
+    }
+    
+    try {
+      String formattedURL = String.format("https://www.googleapis.com/books/v1/volumes?q={%s}&maxResults=%d&country=US",
+        searchTerm, maxResults);
+		  URL url = new URL(formattedURL);
+		  HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+		  conn.setRequestMethod("GET");
+		  conn.setRequestProperty("Accept", "application/json");
+
+      Reader streamReader = null;
+      if (conn.getResponseCode() != 200) {
+        streamReader = new InputStreamReader(conn.getErrorStream());
+      } else {
+        streamReader = new InputStreamReader(conn.getInputStream());
+      }
+      BufferedReader br = new BufferedReader(streamReader);
+
+      String output;
+      while ((output = br.readLine()) != null) {
+        fullOutput += output;
+      }
+
+      if (conn.getResponseCode() != 200) {
+        System.err.println(output);
```

Comment:
How about logging `fullOutput` instead (just in case the output is longer than one line in the event of an API error)?

Reply:
This was my original intention, I just used the wrong variable name. Thanks for catching this!